File: cpp/92. Reverse Linked List II.cpp
/*
2016-9-23 21:08:46

https://leetcode.com/problems/reverse-linked-list-ii/

Reverse a linked list from position m to n. Do it in-place and in one-pass.

For example:
Given 1->2->3->4->5->NULL, m = 2 and n = 4,

return 1->4->3->2->5->NULL.

Note:
Given m, n satisfy the following condition:
1 ≤ m ≤ n ≤ length of list.
*/


//这个题目对链表的理解 考察比较好


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        
		if (head == nullptr || (m == n))
			return head;
		
		ListNode *a = nullptr, *b = nullptr, *c = nullptr, *d = nullptr;
		
		ListNode* pre = nullptr;
		ListNode* cur = nullptr;
		ListNode* nex = nullptr;
		
		int i = 0;
		
		if (m == 1)
		{
			a = nullptr;
			b = head;
			
			++ i;
		}
		else
		{
			a = head;
			b = head->next;
			
			++ i;
			
			while ((++ i) != m)
			{
				a = a->next;
				b = b->next;
			}
		}
		
		pre = b;
		cur = b->next;
		
		while ((++ i) != n)
		{
			nex = cur->next;
			cur->next = pre;
			pre = cur;
			cur = nex;
		}
		
		c = cur;
		d = cur->next;
		
		cur->next = pre;
		
		
		b->next = d;
		
		if (a != nullptr)
		{
			a->next = c;
			
			return head;
		}
		else
		{
			return c;
		}
		
		
    }
	
	
};

File: java/src/main/java/club/cser/leetcode/92. Reverse Linked List II.java
package club.cser.leetcode;

class ReverseLinkedListII {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        // 前面增加一个节点
        ListNode pHead = new ListNode(0, head);

        int count = 0;

        while (count + 1 != m) {
            ++ count;
            pHead = pHead.next;
        }

        // cur 此时指向 m
        ListNode cur = pHead.next;
        // 保存m，反转后m为最后一个，需要为m.next赋值
        ListNode tmpTail = pHead.next;
        // 置为null 因为要用tmp.next = pHead.next;
        pHead.next = null;

        while (cur != null && (count != n)) {
            // cur向后走一步
            ListNode tmp = cur;
            cur = cur.next;
            ++ count;

            //更新结果pHead
            tmp.next = pHead.next;
            pHead.next = tmp;
        }

        // 为m.next赋值
        if (tmpTail != null)
            tmpTail.next = cur;

        if (m == 1)
            return pHead.next;
        else
            return head;

    }
}